core/src/main/java/com/linecorp/armeria/client/cookie/DefaultCookieJar.java

```diff
@@ -0,0 +1,200 @@
+        cookiePolicy = CookiePolicy.ACCEPT_ORIGINAL_SERVER;
+        allCookies = new HashSet<>();
+        hostIndex = new HashMap<>();
+        domainIndex = new HashMap<>();
+        lock = new ReentrantLock();
+    }
+
+    @Override
+    public Cookies get(URI uri) {
+        requireNonNull(uri, "uri");
+        if (allCookies.isEmpty()) {
+            return Cookies.of();
+        }
+        String path = uri.getPath();
+        if (path == null || path.isEmpty()) {
+            path = "/";
+        }
+        final String host = uri.getHost();
+        final boolean secure = "https".equalsIgnoreCase(uri.getScheme());
+        final Set<Cookie> cookies = new HashSet<>();
+        lock.lock();
+        try {
+            getIndex(hostIndex, cookies, path, secure, (h) -> h.equals(host));
+            getIndex(domainIndex, cookies, path, secure, (d) -> CookieJar.domainMatches(d, host));
+        } finally {
+            lock.unlock();
+        }
+        return Cookies.of(cookies);
+    }
+
+    @Override
+    public void set(URI uri, Cookies cookies) {
+        requireNonNull(uri, "uri");
+        requireNonNull(cookies, "cookies");
+        final String host = uri.getHost();
+        lock.lock();
+        try {
+            for (Cookie cookie : cookies) {
+                // If no path is specified, compute the default path
+                // https://tools.ietf.org/html/rfc6265#section-5.1.4
+                if (cookie.path() == null) {
+                    String path = uri.getPath();
+                    if (!path.endsWith("/")) {
+                        final int i = path.lastIndexOf('/');
+                        if (i > 0) {
+                            path = path.substring(0, i + 1);
+                        } else {
+                            path = "/";
+                        }
+                    }
+                    cookie = cookie.toBuilder().path(path).build();
+                }
+                // remove existing cookie if present
+                allCookies.remove(cookie);
+                if (!cookie.isExpired() && cookiePolicy.shouldAccept(uri, cookie)) {
+                    allCookies.add(cookie);
+                    // a cookie always matches its request's host, so it's always in hostIndex
+                    addIndex(hostIndex, host, cookie);
+                    if (cookie.domain() != null) {
+                        addIndex(domainIndex, requireNonNull(cookie.domain()), cookie);
+                    }
+                }
+            }
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    @Override
+    public void setCookiePolicy(CookiePolicy cookiePolicy) {
+        requireNonNull(cookiePolicy, "cookiePolicy");
+        this.cookiePolicy = cookiePolicy;
+    }
+
+    private <T> void getIndex(Map<T, Set<Cookie>> index, Set<Cookie> cookies, String path, boolean secure,
+                              Predicate<T> condition) {
+        for (T key : index.keySet()) {
```

Comment:
Fine to use `entrySet()` so that we don't use `index.get()` again?
```java
for (Entry<T, Set<Cookie>> entry : index.entrySet()) {
    if (condition.test(entry.getKey())) {
        final Set<Cookie> matchedCookies = entry.getValue();
```